### DISCUSSION TITLE
tilt and shift implementation

### DISCUSSION THREAD

**Discussion by elafmusa** [atcollab/at#512] [Python AT]
Hi everyone,

In order to implement shift errors in PyAT i am using:

**at.shift_elem(latticeElement, shiftx, shifty, relative=False)**

And this introduce horizontal and vertical disclaimants of the elements. My question is how to introduce the misalignment in the s-direction for the entry of the beam element called DS in MADX.

Also, for tilt implementation, i am using:

**at.tilt_elem(ring[i], tilt, relative=False)**

As far as i understood this tilt angle corresponds to a rotation of the element looking in the direction of the beam
Corresponding to DPSI in MADX, i want to implement rotations around the x axis and y axis, DPHI and DTHETA, how can i implement it in PYAT?

**Comment by swhite2401** [comment]
Hello @elafmusa,

@simoneliuzzo is the expert of all these manipulations and may have private functions to do these.
However, they are not included in pyAT (yet).

To do a longitudinal shift, it has to be done manually: you may increase the length of the drift upstream and shorten the one downstream or inversely

To do a pitch or a yawn you can use the variables T1 and T2 that change the referential at the entrance and exit of the element, for example to rotate the element around the x-axis by psi:
T1 = [0 ,0, L/2*sin(psi), -psi, 0, 0]
T2 = [0 ,0, L/2*sin(psi), psi, 0, 0]

and similarly action on T[0], T[1] for a rotation around the y axis.

I hope I got this right... @simoneliuzzo please correct me if not.

**Reply by elafmusa** [reply]
Dear @swhite2401 ,

Thank you for the answered,
I have just tried ti implement your suggestion for the x-axis:

psi = 45
for i in quad_indexes:
    a = ring[i].Length/2*sin(psi)
    ring[i].T1[2] = a
    ring[i].T1[3] = -psi
    ring[i].T2[2] = a
    ring[i].T2[3] = psi

I got the error 'Quadrupole' object has no attribute 'T1' this is a problem i faced before when i tried to access R1 and R2, these parameters don't exist unless the element is already shifted or rotated, so i cannot assign values to it, 

I think of assigning a zero shift to the element so that it has T1 and T2 then i can change their values, but i am not sure if it's the best way to do it 

#shift errors

shiftx =0
shifty = 0
for i in quad_indexes:
    at.shift_elem(ring[i], shiftx, shifty, relative=False)

@simoneliuzzo

**Reply by swhite2401** [reply]
Dear @elafmusa,
This is because you are trying to access an attribute that does not exist yet with `ring[i].T1[2]`, you may try:
```
t1 = [0.0, 0.0, a, -psi, 0.0, 0.0]
t2 = [0.0, 0.0, a, psi, 0.0, 0.0]
ring[i].T1 = t1
ring[i].T2 =t2
```

**Reply by elafmusa** [reply]
@swhite2401 I tried to do it this way, but the attribute still doesn't exist when i tried to access it later  

psi = 100*1.e-6

for i in used_quadrpoles_families_ind:
    for j in i:
        a = ring[j].Length/2*sin(psi)
        #print(a)
        t1 = [ a, -psi, a, -psi, 0.0, 0.0]
        #print(t1)
        t2 = [a, psi, a, psi, 0.0, 0.0]
        ring[i].T1 = t1
        print(ring[i].T1)
        ring[i].T2 =t2
AttributeError: 'Lattice' object has no attribute 'T1'

**Reply by swhite2401** [reply]
Sorry I did not look at your code very carefully, in fact I do not understand very well what you are trying to do why do you have this nested loop?

In any case if you want to apply a pitch of `psi `to all quads you can write:

```
quads_refpts = ring.get_refpts(at.Quadrupole)
for e in ring[quads_refpts]:
    a = e.Length/2*numpy.sin(psi)
    t1 = [0.0, 0.0, a, -psi, 0.0, 0.0]
    t2 = [0.0, 0.0, a, psi, 0.0, 0.0]
    e.T1 = t1
    e.T2 = t2
```

**Reply by swhite2401** [reply]
If you can share the script you are running I can try to see where the problem is

**Reply by elafmusa** [reply]
Dear swhite,
The nested loop is because i am using set of sets for quadrupoles families, but i can simplify it for now by using the whole quadrupoles, it works now, just i want to ask about the unit for psi should i transform it to degrees inside the sin function ?:
I am tiring to implement rotation around the x and y axis at the same time, 
psi = 100*1.e-6
element_indexes = get_refpts(ring, elements.Quadrupole)

for i in element_indexes:
    a = ring[i].Length/2*sin(psi)
    #print(a)
    t1 = [ a, -psi, a, -psi, 0.0, 0.0]
    #print(t1)
    t2 = [a, psi, a, psi, 0.0, 0.0]
    ring[i].T1 = t1
    print(ring[i].T1)
    ring[i].T2 =t2

**Reply by swhite2401** [reply]
Ok I see, your implementation cannot work as you cannot set attributes of an array of elements just using `ring[indexes].T1`. You have to loop through individual elements.
The angle is expressed in radians.

Below an example:

```
fam_names = ['QF1*', 'QD2*']
psi= 1.0e-6
for fname in fam_names:
    fam = ring.get_elements(fname)
    for e in fam:
       a = e.Length*numpy.sin(psi)
       t1 = [0.0, 0.0, a, psi, 0.0, 0.0]
       t2 = [0.0, 0.0, a, -psi, 0.0, 0.0]
       e.T1 = t1
       e.T2 = t2
       print(e.T1) 
```

**Reply by elafmusa** [reply]
@swhite2401 Thanks for the information,  i managed to implement it now

**Comment by simoneliuzzo** [comment]
Dear @elafmusa,

As @swhite2401 mentioned I wrote some time ago these functions in matlab AT. 

The longitudinal shift function is here: https://gitlab.esrf.fr/BeamDynamics/ATClusterTools/-/blob/master/errors_and_corrections/errors/atset_s_shift.m

For pitch and yaw:

https://gitlab.esrf.fr/BeamDynamics/ATClusterTools/-/blob/master/errors_and_corrections/errors/atsetpitch.m
https://gitlab.esrf.fr/BeamDynamics/ATClusterTools/-/blob/master/errors_and_corrections/errors/atsetyaw.m

let me know if you can access the files and if you find any issue.

best regards
Simone

**Reply by elafmusa** [reply]
@simoneliuzzo Thank you for the info, i can access the file, i will work on implementing this in my PyAT code and give you an update with the results

**Reply by swhite2401** [reply]
Very nice, in case you have a pyAT implementation of these we should integrate them in the distribution! you may take a look at `at.shift_elem` for code style

**Comment by lfarv** [comment]
@elafmusa, @swhite2401, @simoneliuzzo  :

> t1 = [ a, -psi, a, -psi, 0.0, 0.0]

Be careful: the 2nd and 4th coordinates are not angles, but momenta. So the rotation angle that you get with such a t1 displacement vector depends on the energy of the particle: the angle you get is: x' = p<sub>x</sub>/(1+&delta;).

With this, you introduce a kind of "dispersive kick", rather than a change or referential.

To get a rotation around the x or z axis, one needs a new element which which would increment the 2nd or 4th coordinate by (1+&delta;)*psi.

Note that the Matlab implementation is also wrong.

**Reply by swhite2401** [reply]
Ah thanks @lfarv, right! That is a problem...
If I understand well we need to change the passmethods... although not complicated this is much more involved than I thought...
It is nevertheless a most needed feature, maybe we should post it as an issue to be solved in the near future?

**Reply by swhite2401** [reply]
In fact we should be able to produce the desired transform using a combination of `T` and `R`, I'll try to propose something

**Reply by lfarv** [reply]
> In fact we should be able to produce the desired transform using a combination of T and R, I'll try to propose something

Good idea ! The combination with an s-rotation may be tricky, let's see !

**Reply by simoneliuzzo** [reply]
nice! Luckily pitch and yaw are seldom specified.
